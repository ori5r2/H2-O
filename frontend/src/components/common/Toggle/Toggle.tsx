import { ChangeEventHandler, HTMLAttributes, useId } from 'react';
import { css } from '@emotion/react';
import styled from '@emotion/styled';

interface Props extends HTMLAttributes<HTMLDivElement> {
  isChecked: boolean;
  size: 'small' | 'large';
  handleChangeToggle: ChangeEventHandler<HTMLInputElement>;
}

function Toggle({ isChecked, size, handleChangeToggle, ...restProps }: Props) {
  const id = useId();

  return (
    <ToggleContainer size={size} {...restProps}>
      <input
        type='checkbox'
        id={id}
        onChange={handleChangeToggle}
        css={css`
          display: none;
        `}
      />
      <ToggleLabel htmlFor={id} isChecked={isChecked} size={size}>
        <ToggleSwitch isChecked={isChecked} size={size} />
        <ToggleItem isChecked={isChecked} size={size}>
          외장
        </ToggleItem>
        <ToggleItem isChecked={!isChecked} size={size}>
          내장
        </ToggleItem>
      </ToggleLabel>
    </ToggleContainer>
  );
}

export default Toggle;

const ToggleContainer = styled.div<Pick<Props, 'size'>>`
  ${({ theme }) => theme.flex.flexCenterRow};
  position: relative;
  background-color: ${({ theme }) => theme.colors.white};
  border: 1px solid ${({ theme }) => theme.colors.gray100};
  border-radius: ${({ size }) => (size === 'small' ? 22 : 24)}px;
  width: ${({ size }) => (size === 'small' ? 154 : 213)}px;
  height: ${({ size }) => (size === 'small' ? 44 : 48)}px;
  pointer-events: none;
`;

const ToggleLabel = styled.label<Pick<Props, 'size' | 'isChecked'>>`
  ${({ theme }) => theme.flex.flexCenterRow};
  position: relative;
  overflow: hidden;
  cursor: pointer;
`;

const ToggleSwitch = styled.span<Pick<Props, 'size' | 'isChecked'>>`
  position: absolute;
  left: ${({ isChecked, size }) => (isChecked ? 0 : size === 'small' ? 72 : 101)}px;
  background: ${({ theme }) => theme.colors.primary500};
  border-radius: 20px;
  width: ${({ size }) => (size === 'small' ? 72 : 101)}px;
  height: ${({ size }) => (size === 'small' ? 36 : 40)}px;
  transition: all 0.2s ease-in-out;
`;

const ToggleItem = styled.div<Pick<Props, 'size' | 'isChecked'>>`
  ${({ theme }) => theme.flex.flexCenterRow}
  ${({ theme, isChecked }) => (isChecked ? theme.typography.TextKRMedium14 : theme.typography.TextKRRegular14)}
  color: ${({ theme, isChecked }) => (isChecked ? theme.colors.blueBg : theme.colors.gray400)};
  width: ${({ size }) => (size === 'small' ? 72 : 101)}px;
  height: ${({ size }) => (size === 'small' ? 36 : 40)}px;
  z-index: 1;
  transition: all 0.2s ease-in-out;
  pointer-events: ${({ isChecked }) => (isChecked ? 'none' : 'fill')};
`;
