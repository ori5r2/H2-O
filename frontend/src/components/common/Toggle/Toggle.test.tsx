import Toggle from './Toggle';
import { fireEvent, render, screen } from '@/tests/test-util';

describe('Toggle 컴포넌트 테스트', () => {
  it('Toggle 컴포넌트가 정상적으로 렌더링된다.', () => {
    const handleChangeToggle = vi.fn();
    render(<Toggle data-testid='toggle' size='small' isChecked={true} handleChangeToggle={handleChangeToggle} />);

    expect(screen.getByTestId('toggle')).toBeInTheDocument();
  });

  it('Toggle의 onChange 이벤트가 정상적으로 호출된다.', () => {
    const handleChangeToggle = vi.fn();
    render(<Toggle size='small' isChecked={true} handleChangeToggle={handleChangeToggle} />);

    const toggleInput = screen.getByText('외장');

    fireEvent.click(toggleInput);
    expect(handleChangeToggle).toHaveBeenCalledTimes(1);
  });

  describe('Toggle 스냅샷 테스트', () => {
    it('size 속성이 small, isChecked 속성이 true일 경우', () => {
      const handleChangeToggle = vi.fn();
      render(<Toggle data-testid='toggle' size='small' isChecked={true} handleChangeToggle={handleChangeToggle} />);

      expect(screen.getByTestId('toggle')).toMatchSnapshot();
    });

    it('size 속성이 small, isChecked 속성이 false 경우', () => {
      const handleChangeToggle = vi.fn();
      render(<Toggle data-testid='toggle' size='small' isChecked={false} handleChangeToggle={handleChangeToggle} />);

      expect(screen.getByTestId('toggle')).toMatchSnapshot();
    });

    it('size 속성이 large, isChecked 속성이 true일 경우', () => {
      const handleChangeToggle = vi.fn();
      render(<Toggle data-testid='toggle' size='large' isChecked={true} handleChangeToggle={handleChangeToggle} />);

      expect(screen.getByTestId('toggle')).toMatchSnapshot();
    });

    it('size 속성이 large, isChecked 속성이 false 경우', () => {
      const handleChangeToggle = vi.fn();
      render(<Toggle data-testid='toggle' size='large' isChecked={false} handleChangeToggle={handleChangeToggle} />);

      expect(screen.getByTestId('toggle')).toMatchSnapshot();
    });
  });
});
