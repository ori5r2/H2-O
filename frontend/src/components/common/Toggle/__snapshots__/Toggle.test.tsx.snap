// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle 컴포넌트 테스트 > Toggle 스냅샷 테스트 > size 속성이 large, isChecked 속성이 false 경우 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  background-color: #FFFFFF;
  border: 1px solid #EDEDEE;
  border-radius: 24px;
  width: 213px;
  height: 48px;
  pointer-events: none;
}

.emotion-1 {
  display: none;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.emotion-3 {
  position: absolute;
  left: 101px;
  background: #00397B;
  border-radius: 20px;
  width: 101px;
  height: 40px;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 400;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #B6B8BB;
  width: 101px;
  height: 40px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: fill;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #F4F5F7;
  width: 101px;
  height: 40px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: none;
}

<div
  class="emotion-0"
  data-testid="toggle"
>
  <input
    class="emotion-1"
    id=":r5:"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for=":r5:"
  >
    <span
      class="emotion-3"
    />
    <div
      class="emotion-4"
    >
      외장
    </div>
    <div
      class="emotion-5"
    >
      내장
    </div>
  </label>
</div>
`;

exports[`Toggle 컴포넌트 테스트 > Toggle 스냅샷 테스트 > size 속성이 large, isChecked 속성이 true일 경우 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  background-color: #FFFFFF;
  border: 1px solid #EDEDEE;
  border-radius: 24px;
  width: 213px;
  height: 48px;
  pointer-events: none;
}

.emotion-1 {
  display: none;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.emotion-3 {
  position: absolute;
  left: 0px;
  background: #00397B;
  border-radius: 20px;
  width: 101px;
  height: 40px;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #F4F5F7;
  width: 101px;
  height: 40px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 400;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #B6B8BB;
  width: 101px;
  height: 40px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: fill;
}

<div
  class="emotion-0"
  data-testid="toggle"
>
  <input
    class="emotion-1"
    id=":r4:"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for=":r4:"
  >
    <span
      class="emotion-3"
    />
    <div
      class="emotion-4"
    >
      외장
    </div>
    <div
      class="emotion-5"
    >
      내장
    </div>
  </label>
</div>
`;

exports[`Toggle 컴포넌트 테스트 > Toggle 스냅샷 테스트 > size 속성이 small, isChecked 속성이 false 경우 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  background-color: #FFFFFF;
  border: 1px solid #EDEDEE;
  border-radius: 22px;
  width: 154px;
  height: 44px;
  pointer-events: none;
}

.emotion-1 {
  display: none;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.emotion-3 {
  position: absolute;
  left: 72px;
  background: #00397B;
  border-radius: 20px;
  width: 72px;
  height: 36px;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 400;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #B6B8BB;
  width: 72px;
  height: 36px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: fill;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #F4F5F7;
  width: 72px;
  height: 36px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: none;
}

<div
  class="emotion-0"
  data-testid="toggle"
>
  <input
    class="emotion-1"
    id=":r3:"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for=":r3:"
  >
    <span
      class="emotion-3"
    />
    <div
      class="emotion-4"
    >
      외장
    </div>
    <div
      class="emotion-5"
    >
      내장
    </div>
  </label>
</div>
`;

exports[`Toggle 컴포넌트 테스트 > Toggle 스냅샷 테스트 > size 속성이 small, isChecked 속성이 true일 경우 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  background-color: #FFFFFF;
  border: 1px solid #EDEDEE;
  border-radius: 22px;
  width: 154px;
  height: 44px;
  pointer-events: none;
}

.emotion-1 {
  display: none;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.emotion-3 {
  position: absolute;
  left: 0px;
  background: #00397B;
  border-radius: 20px;
  width: 72px;
  height: 36px;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #F4F5F7;
  width: 72px;
  height: 36px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR;
  font-size: 14px;
  font-weight: 400;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #B6B8BB;
  width: 72px;
  height: 36px;
  z-index: 1;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: fill;
}

<div
  class="emotion-0"
  data-testid="toggle"
>
  <input
    class="emotion-1"
    id=":r2:"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for=":r2:"
  >
    <span
      class="emotion-3"
    />
    <div
      class="emotion-4"
    >
      외장
    </div>
    <div
      class="emotion-5"
    >
      내장
    </div>
  </label>
</div>
`;
